src/System.Management.Automation/engine/hostifaces/PSTask.cs

```diff
@@ -842,6 +888,62 @@ private void CheckForComplete()
             }
         }
 
+        private Runspace GetRunspace(int taskId)
+        {
+            var runspaceName = string.Format(CultureInfo.InvariantCulture, "{0}:{1}", PSTask.RunspaceName, taskId);
+
+            if (_runspacePool.TryDequeue(out Runspace runspace))
+            {
+                if (runspace.RunspaceStateInfo.State == RunspaceState.Opened &&
+                    runspace.RunspaceAvailability == RunspaceAvailability.Available)
+                {
+                    try
+                    {
+                        runspace.ResetRunspaceState();
+                        runspace.Name = runspaceName;
+                        return runspace;
+                    }
+                    catch
+                    {
+                        // If the runspace cannot be reset for any reason, remove it.
+                    }
+                }
+
+                RemoveActiveRunspace(runspace);
+            }
+
+            // Create and initialize a new Runspace
+            var iss = InitialSessionState.CreateDefault2();
+            iss.LanguageMode = (SystemPolicy.GetSystemLockdownPolicy() == SystemEnforcementMode.Enforce)
+                ? PSLanguageMode.ConstrainedLanguage : PSLanguageMode.FullLanguage;
+            runspace = RunspaceFactory.CreateRunspace(iss);
+            runspace.Name = runspaceName;
+            runspace.Open();
+            _activeRunspaces.TryAdd(runspace.Id, runspace);
+
+            return runspace;
+        }
+
+        private void ReturnRunspace(PSTaskBase task)
+        {
+            var runspace = task.Runspace;
+            Dbg.Assert(runspace != null, "Task runspace cannot be null.");
+            if (runspace.RunspaceStateInfo.State == RunspaceState.Opened &&
```

Comment:
Is it important to check the runspace state if we will reset the runspace before use?

Reply:
Yes.  Runspace cannot be reset or used if it is not in the right state.